ament_copyright/ament_copyright/__init__.py

```diff
@@ -12,31 +12,354 @@
+
+
+def scan_past_coding_and_shebang_lines(content):
+    index = 0
+    while (
+        is_comment_line(content, index) and
+        (is_coding_line(content, index) or
+         is_shebang_line(content, index))
+    ):
+        index = get_index_of_next_line(content, index)
+    return index
+
+
+def scan_past_empty_lines(content, index):
+    while is_empty_line(content, index):
+        index = get_index_of_next_line(content, index)
+    return index
+
+
+def get_index_of_next_line(content, index):
+    index_n = content.find('\n', index)
+    index_r = content.find('\r', index)
+    if index_n == -1 and index_r == -1:
+        return len(content)
+    if index_n == -1:
+        return index_r + 1
+    if index_r == -1:
+        return index_n + 1
+    return min(index_n + 1, index_r + 1)
+
+
+def is_comment_line(content, index):
+    return content[index] == '#' or content[index:index + 1] == '//'
+
+
+def is_coding_line(content, index):
+    end_index = get_index_of_next_line(content, index)
+    line = content[index:end_index]
+    return 'coding=' in line or 'coding:' in line
+
+
+def is_shebang_line(content, index):
+    return content[index:index + 1] == '#!'
+
+
+def is_empty_line(content, index):
+    return get_index_of_next_line(content, index) == index + 1
+
+
+def get_comment(filename, msg):
+    if filename.endswith('.py'):
+        line_prefix = '# '
+    else:
+        line_prefix = '// '
+
+    comment = ''
+    index = 0
+    while True:
+        new_index = get_index_of_next_line(msg, index)
+        if new_index == index:
+            break
+        comment += line_prefix + msg[index:new_index]
+        index = new_index
+    return comment
+
 
 def check_file_for_copyright(filename):
-    with open(filename, 'r') as f:
-        lines = f.readlines()
-        comment = ''
-        block_comment = False
-        for line in lines:
-            line = line.strip()
-            # skip empty lines
-            if not line:
-                continue
-            # identify block comments
-            if line.startswith('/*'):
-                block_comment = True
-            # collect all comment lines
-            if block_comment or line.startswith('#') or line.startswith('//'):
-                comment += '%s\n' % line
-                if block_comment and '*/' in line:
-                    break
-            else:
-                break
-        return 'copyright' in comment.lower()
+    with open(filename, 'r') as h:
+        content = h.read()
+
+    index = scan_past_coding_and_shebang_lines(content)
+    index = scan_past_empty_lines(content, index)
+
+    return is_copyright_line(content, index)
+
+
+def is_copyright_line(content, index):
+    if not is_comment_line(content, index):
+        return False
+
+    end_index = get_index_of_next_line(content, index)
+    line = content[index:end_index].lower()
+
+    if line.startswith('#'):
+        line = line[1:]
+    elif line.startswith('//'):
```

Comment:
Same thing. At the very least it might be possible to consolidate this comment related logic.